{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numFeatures 10\n",
      "val acc 0.848\n",
      "train acc 0.8655913978494624\n",
      "y_pred [1. 1. 0. 1. 0. 1. 0. 0. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 0. 0.\n",
      " 0. 1. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 1. 1. 1. 1. 1. 0. 0. 1. 1. 0. 1.\n",
      " 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 1.\n",
      " 0. 0. 0. 1. 1. 1. 1. 0. 0. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 0. 1. 1. 0.\n",
      " 0. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 0. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1.\n",
      " 1. 0. 1. 0. 0.]\n",
      "y_true 110    1.0\n",
      "147    1.0\n",
      "494    0.0\n",
      "325    1.0\n",
      "189    0.0\n",
      "      ... \n",
      "41     1.0\n",
      "382    0.0\n",
      "285    1.0\n",
      "412    0.0\n",
      "289    0.0\n",
      "Name: target, Length: 125, dtype: float64\n",
      "[0.848]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAGdCAYAAAAxCSikAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/TGe4hAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAen0lEQVR4nO3dfUzV5/3/8Rce7rwD7GzxYJmiyZSG6lqMVpAsNQ7nHNYlS7FdtRrsaueCqHWDeNd6A/WmxkQLjdQmc2NqpO1iuhMnS2rnTTuVYdKI1XhXvAGdZAUyUlG8fn8Yzvd3CloPEXhDn4/kJD0frg9c13Vqz7OfcziGOOecAAAADOvV1RMAAAD4LgQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzAvt6gk8LHfu3NHVq1fVv39/hYSEdPV0AADAA3DOqaGhQXFxcerV697XUXpMsFy9elXx8fFdPQ0AANAOly5d0uOPP37Pr/eYYOnfv7+kuwuOiorq4tkAAIAHUV9fr/j4eP/z+L30mGBpeRkoKiqKYAEAoJv5rrdz8KZbAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB57QqWwsJCJSQkKDIyUsnJyTp48OB9x5eUlGj06NHq06ePvF6v5syZo9ra2oAxmzdv1ogRI9S7d2/Fx8dr4cKF+uabb9ozPQAA0MMEHSy7d+9WTk6Oli5dqoqKCqWlpWnKlCmqqqpqc/yhQ4c0a9YsZWVl6eTJk9qzZ4+OHTumuXPn+seUlJQoNzdXK1eu1KlTp7R9+3bt3r1beXl57V8ZAADoMYIOlk2bNikrK0tz585VYmKiNm/erPj4eBUVFbU5/vPPP9fQoUOVnZ2thIQETZgwQa+++qqOHz/uH/PZZ58pNTVVL774ooYOHar09HS98MILAWMAAMD3V1DB0tTUpPLycqWnpwccT09P15EjR9o8JyUlRZcvX5bP55NzTteuXVNpaammTp3qHzNhwgSVl5fr6NGjkqTz58/L5/MFjPm2mzdvqr6+PuAGAAB6ptBgBt+4cUPNzc2KjY0NOB4bG6uampo2z0lJSVFJSYkyMzP1zTff6Pbt25o2bZq2bNniHzNjxgz95z//0YQJE+Sc0+3bt/Xaa68pNzf3nnMpKCjQm2++Gcz0AQBAN9WuN92GhIQE3HfOtTrWorKyUtnZ2VqxYoXKy8u1b98+XbhwQfPmzfOPOXDggNauXavCwkL9+9//1ocffqiPP/5Yq1evvucc8vLyVFdX579dunSpPUsBAADdQFBXWAYOHCiPx9Pqasr169dbXXVpUVBQoNTUVC1ZskSSNGrUKPXt21dpaWlas2aNvF6vli9frpkzZ/rfiPvkk0/qf//7n37zm99o6dKl6tWrdVdFREQoIiIimOkDAIBuKqgrLOHh4UpOTlZZWVnA8bKyMqWkpLR5TmNjY6vg8Hg8ku5embnfGOecfwwAAPj+CuoKiyQtWrRIM2fO1JgxYzR+/Hht27ZNVVVV/pd48vLydOXKFe3YsUOSlJGRoVdeeUVFRUWaPHmyqqurlZOTo7FjxyouLs4/ZtOmTXrqqac0btw4nT17VsuXL9e0adP8cQMAAL6/gg6WzMxM1dbWatWqVaqurlZSUpJ8Pp+GDBkiSaqurg74TJbZs2eroaFBW7du1eLFixUTE6OJEydq3bp1/jHLli1TSEiIli1bpitXrujRRx9VRkaG1q5d+xCWCAAAursQ10Nec6mvr1d0dLTq6uoUFRXV1dMBAAAP4EGfv/m7hAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAPIIFAACYR7AAAADzCBYAAGAewQIAAMwjWAAAgHkECwAAMI9gAQAA5hEsAADAvHYFS2FhoRISEhQZGank5GQdPHjwvuNLSko0evRo9enTR16vV3PmzFFtbW3AmK+//lrz58+X1+tVZGSkEhMT5fP52jM9AADQwwQdLLt371ZOTo6WLl2qiooKpaWlacqUKaqqqmpz/KFDhzRr1ixlZWXp5MmT2rNnj44dO6a5c+f6xzQ1NemnP/2pLl68qNLSUp0+fVrFxcUaPHhw+1cGAAB6jBDnnAvmhHHjxunpp59WUVGR/1hiYqKmT5+ugoKCVuM3btyooqIinTt3zn9sy5YtWr9+vS5duiRJevfdd7VhwwZ9+eWXCgsLa9dC6uvrFR0drbq6OkVFRbXrewAAgM71oM/fQV1haWpqUnl5udLT0wOOp6en68iRI22ek5KSosuXL8vn88k5p2vXrqm0tFRTp071j9m7d6/Gjx+v+fPnKzY2VklJScrPz1dzc/M953Lz5k3V19cH3AAAQM8UVLDcuHFDzc3Nio2NDTgeGxurmpqaNs9JSUlRSUmJMjMzFR4erkGDBikmJkZbtmzxjzl//rxKS0vV3Nwsn8+nZcuW6e2339batWvvOZeCggJFR0f7b/Hx8cEsBQAAdCPtetNtSEhIwH3nXKtjLSorK5Wdna0VK1aovLxc+/bt04ULFzRv3jz/mDt37uixxx7Ttm3blJycrBkzZmjp0qUBLzt9W15enurq6vy3lpeXAABAzxMazOCBAwfK4/G0uppy/fr1VlddWhQUFCg1NVVLliyRJI0aNUp9+/ZVWlqa1qxZI6/XK6/Xq7CwMHk8Hv95iYmJqqmpUVNTk8LDw1t934iICEVERAQzfQAA0E0FdYUlPDxcycnJKisrCzheVlamlJSUNs9pbGxUr16BP6YlTFre75uamqqzZ8/qzp07/jFnzpyR1+ttM1YAAMD3S9AvCS1atEjvvfee3n//fZ06dUoLFy5UVVWV/yWevLw8zZo1yz8+IyNDH374oYqKinT+/HkdPnxY2dnZGjt2rOLi4iRJr732mmpra7VgwQKdOXNGf/vb35Sfn6/58+c/pGUCAIDuLKiXhCQpMzNTtbW1WrVqlaqrq5WUlCSfz6chQ4ZIkqqrqwM+k2X27NlqaGjQ1q1btXjxYsXExGjixIlat26df0x8fLz279+vhQsXatSoURo8eLAWLFigP/zhDw9hiQAAoLsL+nNYrOJzWAAA6H465HNYAAAAugLBAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwr13BUlhYqISEBEVGRio5OVkHDx687/iSkhKNHj1affr0kdfr1Zw5c1RbW9vm2F27dikkJETTp09vz9QAAEAPFHSw7N69Wzk5OVq6dKkqKiqUlpamKVOmqKqqqs3xhw4d0qxZs5SVlaWTJ09qz549OnbsmObOndtq7FdffaXXX39daWlpwa8EAAD0WEEHy6ZNm5SVlaW5c+cqMTFRmzdvVnx8vIqKitoc//nnn2vo0KHKzs5WQkKCJkyYoFdffVXHjx8PGNfc3Kxf//rXevPNNzVs2LD2rQYAAPRIQQVLU1OTysvLlZ6eHnA8PT1dR44cafOclJQUXb58WT6fT845Xbt2TaWlpZo6dWrAuFWrVunRRx9VVlbWA83l5s2bqq+vD7gBAICeKahguXHjhpqbmxUbGxtwPDY2VjU1NW2ek5KSopKSEmVmZio8PFyDBg1STEyMtmzZ4h9z+PBhbd++XcXFxQ88l4KCAkVHR/tv8fHxwSwFAAB0I+16021ISEjAfedcq2MtKisrlZ2drRUrVqi8vFz79u3ThQsXNG/ePElSQ0ODXnrpJRUXF2vgwIEPPIe8vDzV1dX5b5cuXWrPUgAAQDcQGszggQMHyuPxtLqacv369VZXXVoUFBQoNTVVS5YskSSNGjVKffv2VVpamtasWaNr167p4sWLysjI8J9z586du5MLDdXp06c1fPjwVt83IiJCERERwUwfAAB0U0FdYQkPD1dycrLKysoCjpeVlSklJaXNcxobG9WrV+CP8Xg8ku5emRk5cqS++OILnThxwn+bNm2ann32WZ04cYKXegAAQHBXWCRp0aJFmjlzpsaMGaPx48dr27Ztqqqq8r/Ek5eXpytXrmjHjh2SpIyMDL3yyisqKirS5MmTVV1drZycHI0dO1ZxcXGSpKSkpICfERMT0+ZxAADw/RR0sGRmZqq2tlarVq1SdXW1kpKS5PP5NGTIEElSdXV1wGeyzJ49Ww0NDdq6dasWL16smJgYTZw4UevWrXt4qwAAAD1aiHPOdfUkHob6+npFR0errq5OUVFRXT0dAADwAB70+Zu/SwgAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzCNYAACAeQQLAAAwj2ABAADmESwAAMA8ggUAAJhHsAAAAPMIFgAAYB7BAgAAzGtXsBQWFiohIUGRkZFKTk7WwYMH7zu+pKREo0ePVp8+feT1ejVnzhzV1tb6v15cXKy0tDQNGDBAAwYM0KRJk3T06NH2TA0AAPRAQQfL7t27lZOTo6VLl6qiokJpaWmaMmWKqqqq2hx/6NAhzZo1S1lZWTp58qT27NmjY8eOae7cuf4xBw4c0AsvvKBPPvlEn332mX74wx8qPT1dV65caf/KAABAjxHinHPBnDBu3Dg9/fTTKioq8h9LTEzU9OnTVVBQ0Gr8xo0bVVRUpHPnzvmPbdmyRevXr9elS5fa/BnNzc0aMGCAtm7dqlmzZj3QvOrr6xUdHa26ujpFRUUFsyQAANBFHvT5O6grLE1NTSovL1d6enrA8fT0dB05cqTNc1JSUnT58mX5fD4553Tt2jWVlpZq6tSp9/w5jY2NunXrlh555JF7jrl586bq6+sDbgAAoGcKKlhu3Lih5uZmxcbGBhyPjY1VTU1Nm+ekpKSopKREmZmZCg8P16BBgxQTE6MtW7bc8+fk5uZq8ODBmjRp0j3HFBQUKDo62n+Lj48PZikAAKAbadebbkNCQgLuO+daHWtRWVmp7OxsrVixQuXl5dq3b58uXLigefPmtTl+/fr12rlzpz788ENFRkbecw55eXmqq6vz3+718hIAAOj+QoMZPHDgQHk8nlZXU65fv97qqkuLgoICpaamasmSJZKkUaNGqW/fvkpLS9OaNWvk9Xr9Yzdu3Kj8/Hz94x//0KhRo+47l4iICEVERAQzfQAA0E0FdYUlPDxcycnJKisrCzheVlamlJSUNs9pbGxUr16BP8bj8Ui6e2WmxYYNG7R69Wrt27dPY8aMCWZaAACghwvqCoskLVq0SDNnztSYMWM0fvx4bdu2TVVVVf6XePLy8nTlyhXt2LFDkpSRkaFXXnlFRUVFmjx5sqqrq5WTk6OxY8cqLi5O0t2XgZYvX66//OUvGjp0qP8KTr9+/dSvX7+HtVYAANBNBR0smZmZqq2t1apVq1RdXa2kpCT5fD4NGTJEklRdXR3wmSyzZ89WQ0ODtm7dqsWLFysmJkYTJ07UunXr/GMKCwvV1NSkX/3qVwE/a+XKlXrjjTfauTQAANBTBP05LFbxOSwAAHQ/HfI5LAAAAF2BYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOaFdvUEHhbnnCSpvr6+i2cCAAAeVMvzdsvz+L30mGBpaGiQJMXHx3fxTAAAQLAaGhoUHR19z6+HuO9Kmm7izp07unr1qvr376+QkJCunk6Xqq+vV3x8vC5duqSoqKiunk6Pxl53Dva5c7DPnYN9DuScU0NDg+Li4tSr173fqdJjrrD06tVLjz/+eFdPw5SoqCj+MHQS9rpzsM+dg33uHOzz/7nflZUWvOkWAACYR7AAAADzCJYeKCIiQitXrlRERERXT6XHY687B/vcOdjnzsE+t0+PedMtAADoubjCAgAAzCNYAACAeQQLAAAwj2ABAADmESzdRGFhoRISEhQZGank5GQdPHjwvuPfeecdJSYmqnfv3hoxYoR27NjRaszXX3+t+fPny+v1KjIyUomJifL5fB21hG6hI/Z58+bNGjFihHr37q34+HgtXLhQ33zzTUctwbx//vOfysjIUFxcnEJCQvTXv/71O8/59NNPlZycrMjISA0bNkzvvvtuqzEffPCBnnjiCUVEROiJJ57QRx991AGz7z46Yp+Li4uVlpamAQMGaMCAAZo0aZKOHj3aQSvoHjrq3+cWu3btUkhIiKZPn/7wJt1dOZi3a9cuFxYW5oqLi11lZaVbsGCB69u3r/vqq6/aHF9YWOj69+/vdu3a5c6dO+d27tzp+vXr5/bu3esfc/PmTTdmzBj385//3B06dMhdvHjRHTx40J04caKzlmVOR+zzn//8ZxcREeFKSkrchQsX3N///nfn9XpdTk5OZy3LHJ/P55YuXeo++OADJ8l99NFH9x1//vx516dPH7dgwQJXWVnpiouLXVhYmCstLfWPOXLkiPN4PC4/P9+dOnXK5efnu9DQUPf555938Grs6oh9fvHFF90777zjKioq3KlTp9ycOXNcdHS0u3z5cgevxq6O2OcWFy9edIMHD3ZpaWnuueee65gFdCMESzcwduxYN2/evIBjI0eOdLm5uW2OHz9+vHv99dcDji1YsMClpqb67xcVFblhw4a5pqamhz/hbqoj9nn+/Plu4sSJAWMWLVrkJkyY8JBm3b09yH/gf//737uRI0cGHHv11VfdM88847///PPPu5/97GcBYyZPnuxmzJjx0ObanT2sff6227dvu/79+7s//vGPD2Oa3d7D3Ofbt2+71NRU995777mXX36ZYHHO8ZKQcU1NTSovL1d6enrA8fT0dB05cqTNc27evKnIyMiAY71799bRo0d169YtSdLevXs1fvx4zZ8/X7GxsUpKSlJ+fr6am5s7ZiHGddQ+T5gwQeXl5f7L5ufPn5fP59PUqVM7YBU902effdbqcZk8ebKOHz/u3+d7jbnXY4fWHmSfv62xsVG3bt3SI4880hlT7BEedJ9XrVqlRx99VFlZWZ09RbMIFuNu3Lih5uZmxcbGBhyPjY1VTU1Nm+dMnjxZ7733nsrLy+Wc0/Hjx/X+++/r1q1bunHjhqS7T5ylpaVqbm6Wz+fTsmXL9Pbbb2vt2rUdviaLOmqfZ8yYodWrV2vChAkKCwvT8OHD9eyzzyo3N7fD19RT1NTUtPm43L5927/P9xpzr8cOrT3IPn9bbm6uBg8erEmTJnXGFHuEB9nnw4cPa/v27SouLu6KKZrVY/625p4uJCQk4L5zrtWxFsuXL1dNTY2eeeYZOecUGxur2bNna/369fJ4PJKkO3fu6LHHHtO2bdvk8XiUnJysq1evasOGDVqxYkWHr8eqh73PBw4c0Nq1a1VYWKhx48bp7NmzWrBggbxer5YvX97h6+kp2npcvn08mMcObXuQfW6xfv167dy5UwcOHGh1pRH3d799bmho0EsvvaTi4mINHDiwK6ZnFldYjBs4cKA8Hk+r/1O8fv16q0pv0bt3b73//vtqbGzUxYsXVVVVpaFDh6p///7+PwBer1c/+tGP/E+skpSYmKiamho1NTV13IKM6qh9Xr58uWbOnKm5c+fqySef1C9/+Uvl5+eroKBAd+7c6fB19QSDBg1q83EJDQ3VD37wg/uOuddjh9YeZJ9bbNy4Ufn5+dq/f79GjRrVmdPs9r5rn8+dO6eLFy8qIyNDoaGhCg0N1Y4dO7R3716Fhobq3LlzXTTzrkewGBceHq7k5GSVlZUFHC8rK1NKSsp9zw0LC9Pjjz8uj8ejXbt26Re/+IV69br7kKempurs2bMBT5pnzpyR1+tVeHj4w1+IcR21z42Njf5/buHxeOTuvuH94S6ihxo/fnyrx2X//v0aM2aMwsLC7jvmux47/J8H2WdJ2rBhg1avXq19+/ZpzJgxnT3Nbu+79nnkyJH64osvdOLECf9t2rRpevbZZ3XixAnFx8d30cwN6Jr3+iIYLb9uu337dldZWelycnJc37593cWLF51zzuXm5rqZM2f6x58+fdr96U9/cmfOnHH/+te/XGZmpnvkkUfchQsX/GOqqqpcv3793O9+9zt3+vRp9/HHH7vHHnvMrVmzprOXZ0ZH7PPKlStd//793c6dO9358+fd/v373fDhw93zzz/f2cszo6GhwVVUVLiKigonyW3atMlVVFT4f3382/vc8mugCxcudJWVlW779u2tfg308OHDzuPxuLfeesudOnXKvfXWW9/7X2vuiH1et26dCw8Pd6Wlpa66utp/a2ho6PT1WdER+/xt/JbQXQRLN/HOO++4IUOGuPDwcPf000+7Tz/91P+1l19+2f3kJz/x36+srHQ//vGPXe/evV1UVJR77rnn3Jdfftnqex45csSNGzfORUREuGHDhrm1a9e627dvd8ZyzHrY+3zr1i33xhtvuOHDh7vIyEgXHx/vfvvb37r//ve/nbQiez755BMnqdXt5Zdfds613mfnnDtw4IB76qmnXHh4uBs6dKgrKipq9X337NnjRowY4cLCwtzIkSPdBx980Amrsasj9nnIkCFtfs+VK1d2zqIM6qh/n/9/BMtdIc5xXRoAANjGe1gAAIB5BAsAADCPYAEAAOYRLAAAwDyCBQAAmEewAAAA8wgWAABgHsECAADMI1gAAIB5BAsAADCPYAEAAOYRLAAAwLz/B8zf5oI5crXiAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from parsingGrowthCurveFiles import *\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import *\n",
    "from sklearn.neural_network import *\n",
    "# import shap\n",
    "import os\n",
    "# import xgboost as xgb\n",
    "\n",
    "\n",
    "debug = False\n",
    "def createDataset(targets,alignedSnpsAsDict,alignedSnpIndexes, outFileName, significantSnpData, onlySigSnps=True, onlyLastTimePoint=False):\n",
    "\tdata = []\n",
    "\tif onlySigSnps:\n",
    "\t\tindexInSnpGenomeOfSigSnps = []\n",
    "\t\tsigSnpGenomePoses = getSignificantSnpIndexes(significantSnpData)\n",
    "\t\tsigSnpGenomePoses.sort(key=lambda a: a[0] + a[1])\n",
    "\t\tindexOfNextSigSnpGenomePos = 0\n",
    "\t\ti = 0\n",
    "\t\tsigSnpGenomePosesAdded = []\n",
    "\t\tfor genomePos in alignedSnpIndexes:\n",
    "\t\t\tif indexOfNextSigSnpGenomePos == len(sigSnpGenomePoses):\n",
    "\t\t\t\tbreak\n",
    "\t\t\tif int(genomePos) == sigSnpGenomePoses[indexOfNextSigSnpGenomePos][0] + sigSnpGenomePoses[indexOfNextSigSnpGenomePos][1]:# sigSnpGenomePoses[max(indexOfNextSigSnpGenomePos-20,0):indexOfNextSigSnpGenomePos+20]:\n",
    "\t\t\t\tsigSnpGenomePosesAdded.append(\"geneStart_\" + str(sigSnpGenomePoses[indexOfNextSigSnpGenomePos][0]) + \"Index_\" +  str(sigSnpGenomePoses[indexOfNextSigSnpGenomePos][1]))\n",
    "\t\t\t\tindexOfNextSigSnpGenomePos += 1\n",
    "\t\t\t\tindexInSnpGenomeOfSigSnps.append(i)\n",
    "\t\t\t\t\n",
    "\t\t\telif int(genomePos) > sigSnpGenomePoses[indexOfNextSigSnpGenomePos][0] + sigSnpGenomePoses[indexOfNextSigSnpGenomePos][1]: # skip if didn't find it\n",
    "\t\t\t\tindexOfNextSigSnpGenomePos += 1\n",
    "\t\t\ti += 1\n",
    "\t\tif indexOfNextSigSnpGenomePos != len(sigSnpGenomePoses):\n",
    "\t\t\t# raise Exception(\"problem\")\n",
    "\t\t\tprint(indexOfNextSigSnpGenomePos, \"idex of next snp isn't len(sigSnpGenomePoses)\", len(sigSnpGenomePoses))\n",
    "\t\t\n",
    "\t\theader = \"\\t\".join([\"fileName\", \"target\"] + [str(idx) for idx in sigSnpGenomePosesAdded])\n",
    "\t\ti = -1\n",
    "\t\tif debug:\n",
    "\t\t\tprint(\"targets keys\", targets.keys())\n",
    "\t\t\t\n",
    "\t\tfor key, val in alignedSnpsAsDict.items():\n",
    "\t\t\ti += 1\n",
    "\t\t\tsnpsToAdd = []\n",
    "\t\t\tval = val.strip()\n",
    "\t\t\tfor idx in indexInSnpGenomeOfSigSnps:\n",
    "\t\t\t\t# print(\"snps\", val[idx])\n",
    "\t\t\t\tsnpsToAdd.append(val[idx])\n",
    "\t\t\ttry:\n",
    "\t\t\t\tdata.append(key + \"\\t\" + str(targets[key]) + \"\\t\" + \"\\t\".join(\n",
    "\t\t\t\tsnpsToAdd)) \n",
    "\t\t\t\t\n",
    "\t\t\texcept KeyError:\n",
    "\t\t\t\tif debug:\n",
    "\t\t\t\t\tprint(\"didn't find\", key)\n",
    "\telse:\n",
    "\t\theader = \"\\t\".join([\"fileName\", \"target\"] + [str(idx) for idx in alignedSnpIndexes])\n",
    "\t\t\n",
    "\t\tfor key,val in alignedSnpsAsDict.items():\n",
    "\t\t\ttry:\n",
    "\t\t\t\tdata.append(key + \"\\t\" + str(targets[key]) + \"\\t\" + \"\\t\".join(\n",
    "\t\t\t\tsnpsToAdd)) \n",
    "\t\t\t\t# need to only consider significant aleles but maybe at firsta can just do all of them\n",
    "\t\t\texcept KeyError:\n",
    "\t\t\t\tif debug:\n",
    "\t\t\t\t\tprint(\"didn't find\",key)\n",
    "\t\t\n",
    "\twith open(outFileName, \"w\") as outFile:\n",
    "\t\toutFile.write(header + \"\\n\")\n",
    "\t\tfor line in data:\n",
    "\t\t\toutFile.write(line + \"\\n\")\n",
    "\n",
    "def trainAndRunShap(dataset): # pandas\n",
    "\tdataset.pop(\"fileName\")\n",
    "\ty = dataset.pop(\"target\").astype(float)\n",
    "\t\n",
    "\tX:pd.DataFrame = pd.get_dummies(dataset)\n",
    "\tvalAccs = []\n",
    "\tfor numFeatures in range(10,11):\n",
    "\t\tprint(\"numFeatures\", numFeatures)\n",
    "\t\treducedFeatures = ['geneStart_2848875Index_1151_T','geneStart_46478Index_-16_A', 'geneStart_46478Index_-16_G', 'geneStart_4617494Index_1329_G', 'geneStart_4617494Index_1329_T', 'geneStart_2848875Index_1151_G', 'geneStart_273987Index_-2105_A', 'geneStart_273987Index_-2105_G', 'geneStart_46478Index_-138_A', 'geneStart_46478Index_-141_C', 'geneStart_46478Index_-138_G', 'geneStart_46478Index_-141_T', 'geneStart_1611762Index_2781_C', 'geneStart_1611762Index_2781_T', 'geneStart_798261Index_1209_G', 'geneStart_1611762Index_2783_G', 'geneStart_4713593Index_1990_A', 'geneStart_798261Index_1203_T', 'geneStart_4713593Index_1994_G', 'geneStart_4713593Index_1994_T', 'geneStart_1611762Index_2780_T', 'geneStart_798261Index_1209_A', 'geneStart_1981205Index_1164_A', 'geneStart_1611762Index_2783_A', 'geneStart_4713593Index_1987_C', 'geneStart_4713593Index_1990_-', 'geneStart_4740524Index_610_G', 'geneStart_1611762Index_2784_G', 'geneStart_1611762Index_2780_C', 'geneStart_798261Index_1203_C', 'geneStart_1610901Index_60_A', 'geneStart_1611762Index_2784_A', 'geneStart_1981205Index_1164_T', 'geneStart_4624378Index_-37_C', 'geneStart_3077707Index_-227_G', 'geneStart_4713593Index_1987_T', 'geneStart_4713593Index_912_G', 'geneStart_4200127Index_909_G', 'geneStart_2571573Index_318_G', 'geneStart_4740524Index_610_C', 'geneStart_4715665Index_-35_C', 'geneStart_4624378Index_-37_T', 'geneStart_4200127Index_909_T', 'geneStart_511450Index_180_A', 'geneStart_1269572Index_1062_G', 'geneStart_4715665Index_-35_A', 'geneStart_511450Index_180_G', 'geneStart_4713593Index_1410_A', 'geneStart_3077707Index_-227_A', 'geneStart_1610901Index_60_T', 'geneStart_1981205Index_1169_A', 'geneStart_1981205Index_1169_C', 'geneStart_4713593Index_912_A', 'geneStart_4717102Index_1014_T', 'geneStart_4717102Index_1014_A', 'geneStart_1269572Index_1062_A', 'geneStart_2571573Index_321_G', 'geneStart_4713593Index_1414_A', 'geneStart_4690801Index_272_C', 'geneStart_4713593Index_1944_T', 'geneStart_1981205Index_1168_T', 'geneStart_133954Index_1904_C', 'geneStart_748777Index_587_C', 'geneStart_1981205Index_223_G', 'geneStart_4690801Index_272_T', 'geneStart_2571573Index_339_T', 'geneStart_2136000Index_275_C', 'geneStart_1981205Index_1168_C', 'geneStart_4713593Index_1414_G', 'geneStart_748777Index_587_T', 'geneStart_2571573Index_349_T', 'geneStart_2571573Index_339_G']\n",
    "\t\tX_reduced = X[reducedFeatures[:numFeatures]]\n",
    "\n",
    "\t\ttrain_X, val_X, train_y, val_y = train_test_split(X_reduced, y, random_state=1)\n",
    "\t\t# Xd = xgboost.DMatrix(X, label=y)\n",
    "\t\t# model = xgboost.train({\"eta\": 1, \"max_depth\": 3, \"base_score\": 0, \"lambda\": 0}, Xd, 1)\n",
    "\n",
    "\t\t# print(\"Model error =\", np.linalg.norm(y - model.predict(Xd)))\n",
    "\t\t# print(model.get_dump(with_stats=True)[0])\n",
    "\t\trfClass = RandomForestClassifier(n_estimators=1000).fit(train_X,train_y)\n",
    "\t\trowToShow = 6\n",
    "\t\t# explainer = shap.TreeExplainer(rfClass)\n",
    "\t\t# Calculate Shap values\n",
    "\t\t# shap_values = explainer.shap_values(val_X)\n",
    "\n",
    "\t\t# vals= np.abs(shap_values).mean(0)\n",
    "\n",
    "\t\t# feature_importance = pd.DataFrame(list(zip(X.columns, shap_values)), columns=['col_name','feature_importance_vals'])\n",
    "\t\t# feature_importance.sort_values(by=['feature_importance_vals'], ascending=False,inplace=True)\n",
    "\t\t# feature_importance.head()\n",
    "\t\t# if debug:\n",
    "\t\t\t# print(\"Expected Value:\", explainer.expected_value)\n",
    "\t\t# \tprint(\"SHAP Values:\", shap_values)\n",
    "\t\t# \tprint(\"sv\",len(shap_values))\n",
    "\t\tvalAcc = (rfClass.predict(val_X) == val_y).mean()\n",
    "\t\tprint(\"val acc\", valAcc)\n",
    "\t\tprint(\"train acc\", (rfClass.predict(train_X) == train_y).mean())\n",
    "\t\t# print(\"val MAE\", sum(abs(rfClass.predict(val_X) - val_y)) / len(val_y))\n",
    "\t\tvalAccs.append(valAcc)\n",
    "\t\t# print(\"train MAE\", sum(abs(rfClass.predict(train_X) - train_y)) / len(train_y))\n",
    "\t\tprint(\"y_pred\", rfClass.predict(val_X))\n",
    "\t\tprint(\"y_true\", val_y)\n",
    "\t\tprint(valAccs)\n",
    "\tplt.plot([i for i in range(1,len(valAccs)+1)],valAccs)\n",
    "\tplt.show()\n",
    "\t# shap.initjs()\n",
    "\t# # shap.force_plot(explainer.expected_value[1], shap_values[:,1], data_for_prediction)\n",
    "\t# shap.summary_plot(shap_values,val_X, max_display=200)\n",
    "\t# plt.savefig(\"summaryPlot.png\", format=\"png\")\n",
    "\t# plt.close()\n",
    "\treturn rfClass\n",
    "\t\n",
    "def getSignificantSnpIndexes(significantSnpData, pValueCutoff=1e-20):\n",
    "\tidxs = []\n",
    "\tfirstLine = True\n",
    "\tfor line in significantSnpData:\n",
    "\t\tif firstLine:\n",
    "\t\t\tfirstLine = False\n",
    "\t\t\tcontinue\n",
    "\t\tcols = line.split('\\t')\n",
    "\t\tgeneStartPos = int(cols[11])\n",
    "\t\tunCorrectedIndexWithinGene = int(cols[2])\n",
    "\t\tsnpName = cols[4] + cols[2]\n",
    "\t\tpvalue = float(cols[0])\n",
    "\t\tif pvalue < pValueCutoff:\n",
    "\t\t\tif debug:\n",
    "\t\t\t\tprint(pvalue,geneStartPos + unCorrectedIndexWithinGene)\n",
    "\t\t\tidxs.append([geneStartPos, unCorrectedIndexWithinGene])\n",
    "\treturn idxs\n",
    "if __name__ == \"__main__\":\n",
    "\t\n",
    "\talignedSnpsPath = \"/Users/cazcullimore/dev/ericksonLabCode/ScaffoldFromAssembly/bovineSelfScaffold.afa\"\n",
    "\talignedIndexPath = \"/Users/cazcullimore/dev/ericksonLabCode/ScaffoldFromAssembly/bovineSelfScaffoldIndexes.txt\"\n",
    "\n",
    "\ttargets = {key + \".vcf\": 1 if val[1] == \"pathogen\" else 0 for key,val in readMetaDataAsDict(\"/Users/cazcullimore/dev/data/metaDataForMetaCatsWithExtraMastitisWithoutScaffoldNames.tsv\").items()}\n",
    "\t# print(len(list(targets.keys())))\n",
    "\tdatasetPath = \"/Users/cazcullimore/dev/data/dataForRFMastitisCommensal.tsv\"\n",
    "\t# if not os.path.exists(datasetPath):\n",
    "\twith open(\"/Users/cazcullimore/dev/ericksonLabCode/ScaffoldFromAssembly/snpsSortedBySignificanceWithGenesContainingThembovineSelfScaffoldPathogenicity.tsv\") as file:\n",
    "\t\tsigSnpData = file.readlines()\n",
    "\t\tcreateDataset(targets, readInFastaAsDict(alignedSnpsPath), readIndexes(alignedIndexPath), datasetPath, significantSnpData=sigSnpData, onlyLastTimePoint=False)\n",
    "\t# else:\n",
    "\t# \tprint(\"found dataset file using that\")\n",
    "\trfClass = trainAndRunShap(pd.read_csv(datasetPath, delimiter='\\t'))\n",
    "\t\"\"\"50 features \n",
    "\tval MSE 0.0024660483365868285\n",
    "train MSE 0.00045887003507715267\n",
    "val MAE 0.024247464402395016\n",
    "train MAE 0.009662864379214084\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rfClass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "explainer = shap.TreeExplainer(rfClass)\n",
    "dataset = pd.read_csv(datasetPath, delimiter='\\t')\n",
    "dataset.pop(\"fileName\")\n",
    "y = dataset.pop(\"target\").astype(float)\n",
    "\n",
    "X:pd.DataFrame = pd.get_dummies(dataset)\n",
    "valAcc = []\n",
    "for numFeatures in range(1,2):\n",
    "    print(\"numFeatures\", numFeatures)\n",
    "    X_reduced = X#[reducedFeatures[:numFeatures]]\n",
    "\n",
    "    train_X, val_X, train_y, val_y = train_test_split(X_reduced, y, random_state=1)\n",
    "# Calculate Shap values\n",
    "shap_values = explainer.shap_values(val_X)\n",
    "\n",
    "# vals= np.abs(shap_values).mean(0)\n",
    "\n",
    "# feature_importance = pd.DataFrame(list(zip(X.columns, shap_values)), columns=['col_name','feature_importance_vals'])\n",
    "# feature_importance.sort_values(by=['feature_importance_vals'], ascending=False,inplace=True)\n",
    "# feature_importance.head()\n",
    "if debug:\n",
    "    print(\"Expected Value:\", explainer.expected_value)\n",
    "    print(\"SHAP Values:\", shap_values)\n",
    "    print(\"sv\",len(shap_values))\n",
    "# print(\"val acc\", (rfClass.predict(val_X) == val_y).mean())\n",
    "# print(\"train acc\", (rfClass.predict(train_X) == train_y).mean())\n",
    "print(\"val MSE\", sum((rfClass.predict(val_X) - val_y)**2) / len(val_y))\n",
    "print(\"train MSE\", sum((rfClass.predict(train_X) - train_y)**2) / len(train_y))\n",
    "print(\"val MAE\", sum(abs(rfClass.predict(val_X) - val_y)) / len(val_y))\n",
    "print(\"train MAE\", sum(abs(rfClass.predict(train_X) - train_y)) / len(train_y))\n",
    "print(\"y_pred\", rfClass.predict(val_X))\n",
    "print(\"y_true\", val_y)\n",
    "shap.initjs()\n",
    "# shap.force_plot(explainer.expected_value[1], shap_values[:,1], data_for_prediction)\n",
    "shap.summary_plot(shap_values,val_X, max_display=200)\n",
    "plt.savefig(\"summaryPlot.png\", format=\"png\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shap shape (125, 2051)\n",
      "val shape (125, 2051)\n",
      "(2051,)\n",
      "['geneStart_4616454Index_675_G', 'geneStart_4621757Index_796_A', 'geneStart_4616454Index_675_A', 'geneStart_4621757Index_796_G', 'geneStart_4617494Index_567_G', 'geneStart_4617494Index_567_A', 'geneStart_2848875Index_1151_T', 'geneStart_4619863Index_21_C', 'geneStart_4619863Index_21_G', 'geneStart_46478Index_-16_A', 'geneStart_46478Index_-16_G', 'geneStart_4617494Index_1329_G', 'geneStart_4617494Index_1329_T', 'geneStart_2848875Index_1151_G', 'geneStart_273987Index_-2105_A', 'geneStart_273987Index_-2105_G', 'geneStart_46478Index_-138_A', 'geneStart_46478Index_-141_C', 'geneStart_46478Index_-138_G', 'geneStart_46478Index_-141_T', 'geneStart_1611762Index_2781_C', 'geneStart_1611762Index_2781_T', 'geneStart_798261Index_1209_G', 'geneStart_1611762Index_2783_G', 'geneStart_4713593Index_1990_A', 'geneStart_798261Index_1203_T', 'geneStart_4713593Index_1994_G', 'geneStart_4713593Index_1994_T', 'geneStart_1611762Index_2780_T', 'geneStart_798261Index_1209_A', 'geneStart_1981205Index_1164_A', 'geneStart_1611762Index_2783_A', 'geneStart_4713593Index_1987_C', 'geneStart_4713593Index_1990_-', 'geneStart_4740524Index_610_G', 'geneStart_1611762Index_2784_G', 'geneStart_1611762Index_2780_C', 'geneStart_798261Index_1203_C', 'geneStart_1610901Index_60_A', 'geneStart_1611762Index_2784_A', 'geneStart_1981205Index_1164_T', 'geneStart_4624378Index_-37_C', 'geneStart_3077707Index_-227_G', 'geneStart_4713593Index_1987_T', 'geneStart_4713593Index_912_G', 'geneStart_4200127Index_909_G', 'geneStart_2571573Index_318_G', 'geneStart_4740524Index_610_C', 'geneStart_4715665Index_-35_C', 'geneStart_4624378Index_-37_T', 'geneStart_4200127Index_909_T', 'geneStart_511450Index_180_A', 'geneStart_1269572Index_1062_G', 'geneStart_4715665Index_-35_A', 'geneStart_511450Index_180_G', 'geneStart_4713593Index_1410_A', 'geneStart_3077707Index_-227_A', 'geneStart_1610901Index_60_T', 'geneStart_1981205Index_1169_A', 'geneStart_1981205Index_1169_C', 'geneStart_4713593Index_912_A', 'geneStart_4717102Index_1014_T', 'geneStart_4717102Index_1014_A', 'geneStart_1269572Index_1062_A', 'geneStart_2571573Index_321_G', 'geneStart_4713593Index_1414_A', 'geneStart_4690801Index_272_C', 'geneStart_4713593Index_1944_T', 'geneStart_1981205Index_1168_T', 'geneStart_133954Index_1904_C', 'geneStart_748777Index_587_C', 'geneStart_1981205Index_223_G', 'geneStart_4690801Index_272_T', 'geneStart_2571573Index_339_T', 'geneStart_2136000Index_275_C', 'geneStart_1981205Index_1168_C', 'geneStart_4713593Index_1414_G', 'geneStart_748777Index_587_T', 'geneStart_2571573Index_349_T', 'geneStart_2571573Index_339_G']\n"
     ]
    }
   ],
   "source": [
    "print(\"shap shape\",shap_values[1].shape)\n",
    "print(\"val shape\",val_X.shape)\n",
    "print(abs(shap_values[1]).sum(0).shape)\n",
    "val_X.head()\n",
    "namesAndShap = [(name,val) for name, val in zip(val_X.columns, abs(shap_values[1]).sum(0))]\n",
    "namesAndShap.sort(key=lambda a: -a[1])\n",
    "numFeaturesToGet = 80\n",
    "print([name for name, val in namesAndShap][:numFeaturesToGet])\n",
    "# print(namesAndShap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "geneStart_2848875Index_1151_T gadB 1151\n",
      "geneStart_46478Index_-16_A NearestGeneIs:murI -16\n",
      "geneStart_46478Index_-16_G NearestGeneIs:murI -16\n",
      "geneStart_4617494Index_1329_G fecA 1329\n",
      "geneStart_4617494Index_1329_T fecA 1329\n",
      "geneStart_2848875Index_1151_G gadB 1151\n",
      "geneStart_273987Index_-2105_A NearestGeneIs:nanR_1 -2105\n",
      "geneStart_273987Index_-2105_G NearestGeneIs:nanR_1 -2105\n",
      "geneStart_46478Index_-138_A NearestGeneIs:murI -138\n",
      "geneStart_46478Index_-141_C NearestGeneIs:murI -141\n"
     ]
    }
   ],
   "source": [
    "features = ['geneStart_2848875Index_1151_T','geneStart_46478Index_-16_A', 'geneStart_46478Index_-16_G', 'geneStart_4617494Index_1329_G', 'geneStart_4617494Index_1329_T', 'geneStart_2848875Index_1151_G', 'geneStart_273987Index_-2105_A', 'geneStart_273987Index_-2105_G', 'geneStart_46478Index_-138_A','geneStart_46478Index_-141_C', 'geneStart_46478Index_-138_G', 'geneStart_46478Index_-141_T', 'geneStart_1611762Index_2781_C', 'geneStart_1611762Index_2781_T', 'geneStart_798261Index_1209_G', 'geneStart_1611762Index_2783_G', 'geneStart_4713593Index_1990_A', 'geneStart_798261Index_1203_T', 'geneStart_4713593Index_1994_G', 'geneStart_4713593Index_1994_T', 'geneStart_1611762Index_2780_T', 'geneStart_798261Index_1209_A', 'geneStart_1981205Index_1164_A', 'geneStart_1611762Index_2783_A', 'geneStart_4713593Index_1987_C', 'geneStart_4713593Index_1990_-', 'geneStart_4740524Index_610_G', 'geneStart_1611762Index_2784_G', 'geneStart_1611762Index_2780_C', 'geneStart_798261Index_1203_C', 'geneStart_1610901Index_60_A', 'geneStart_1611762Index_2784_A', 'geneStart_1981205Index_1164_T', 'geneStart_4624378Index_-37_C', 'geneStart_3077707Index_-227_G', 'geneStart_4713593Index_1987_T', 'geneStart_4713593Index_912_G', 'geneStart_4200127Index_909_G', 'geneStart_2571573Index_318_G', 'geneStart_4740524Index_610_C', 'geneStart_4715665Index_-35_C', 'geneStart_4624378Index_-37_T', 'geneStart_4200127Index_909_T', 'geneStart_511450Index_180_A', 'geneStart_1269572Index_1062_G', 'geneStart_4715665Index_-35_A', 'geneStart_511450Index_180_G', 'geneStart_4713593Index_1410_A', 'geneStart_3077707Index_-227_A', 'geneStart_1610901Index_60_T', 'geneStart_1981205Index_1169_A', 'geneStart_1981205Index_1169_C', 'geneStart_4713593Index_912_A', 'geneStart_4717102Index_1014_T', 'geneStart_4717102Index_1014_A', 'geneStart_1269572Index_1062_A', 'geneStart_2571573Index_321_G', 'geneStart_4713593Index_1414_A', 'geneStart_4690801Index_272_C', 'geneStart_4713593Index_1944_T', 'geneStart_1981205Index_1168_T', 'geneStart_133954Index_1904_C', 'geneStart_748777Index_587_C', 'geneStart_1981205Index_223_G', 'geneStart_4690801Index_272_T', 'geneStart_2571573Index_339_T', 'geneStart_2136000Index_275_C', 'geneStart_1981205Index_1168_C', 'geneStart_4713593Index_1414_G', 'geneStart_748777Index_587_T', 'geneStart_2571573Index_349_T', 'geneStart_2571573Index_339_G']\n",
    "def getSignificantSnpNames(significantSnpData, featureNames, pValueCutoff=1e-20):\n",
    "\tidxs = []\n",
    "\t\n",
    "\tfor name in featureNames:\n",
    "\t\tfirstLine = True\n",
    "\t\tfor line in significantSnpData:\n",
    "\t\t\tif firstLine:\n",
    "\t\t\t\tfirstLine = False\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tcols = line.split('\\t')\n",
    "\t\t\tgeneStartPos = int(cols[11])\n",
    "\t\t\tunCorrectedIndexWithinGene = int(cols[2])\n",
    "\t\t\tsnpName = cols[4] + cols[2]\n",
    "\t\t\tpvalue = float(cols[0])\n",
    "\t\t\tif pvalue < pValueCutoff:\n",
    "\t\t\t\t\tif str(geneStartPos) in name and cols[2] in name:\n",
    "\t\t\t\t\t\tprint(name, cols[4], cols[2])\n",
    "\n",
    "with open(\"/Users/cazcullimore/dev/ericksonLabCode/ScaffoldFromAssembly/snpsSortedBySignificanceWithGenesContainingThembovineSelfScaffoldPathogenicity.tsv\") as file:\n",
    "\t\tsigSnpData = file.readlines()\n",
    "\t\tgetSignificantSnpNames(significantSnpData=sigSnpData, featureNames=features[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(n_estimators=1000)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(n_estimators=1000)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(n_estimators=1000)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
